File: java/src/Control/Controller.java
package Control;

import GPSUtils.FixType;
import GPSUtils.GPSposition;
import GPSUtils.GPXmaker;
import Sentences.*;

import java.util.ArrayList;
import java.util.Calendar;

/**
 * @author Siôn Griffiths - sig2@aber.ac.uk
 *         Date: 09/03/14
 *         Time: 12:14
 *
 * The Controller class deefines the main logical steps for the NMEA parsing project.
 * It facilitates the creation of all other classes and funcitons as the main point of entry to the system.
 */
public class Controller {

  /**
   * Stream instances to hold stream data
   */
  private final Stream s1;
  private final Stream s2;
  /**
   * Arraylist to hold corrected gps positions
   */
  private final ArrayList<GPSposition> gps;
  /**
   * Calendar instance to hold time of last corrected gps fix
   */
  private Calendar lastFixTime;

  /**
   * Constructs a Controller.
   * Initialises 2 stream instances with stream datafiles.
   */
  public Controller(){
    s1 = new Stream("dataFiles/gps_1.dat");
    s2 = new Stream("dataFiles/gps_2.dat");
    gps = new ArrayList<GPSposition>();
  }

  /**
   * Method to run the NMEA parser. Entry point of the system.
   */
  public void run(){
    while(!(s1.isEndOfStream())){
      syncStream(s1,s2);
    }
    System.out.println("Streams parsed.");
    System.out.println("gps fixes : " + gps.size());
    for(GPSposition g : gps){
      System.out.println(g.toString());
    }
    makeGPX("dataFiles/output.gpx");
  }

  /**
   * Parses a line in a stream file that corresponds to an NMEA sentence.
   * Splits the sentence and creates a new sentence type object corresponding
   * the the sentence being parsed.
   * @param stream the stream currently being parsed
   */
  public void parseSentence(Stream stream){
    String input;
    if((input=stream.getNext()) != null){
      String[] parts = input.split("[,*]");
      stream.count++;
      String type = parts[0];


      switch(type){
        case "$GPGSA" :
          new GSA(parts, stream);
          break;

        case "$GPGGA" :
          new GGA(parts, stream);

          break;

        case "$GPRMC" :
          new RMC(parts, stream);

          break;

        case "$GPGSV" :
          new GSV(parts, stream);
          break;

        case "$GPZDA" :
          new ZDA(parts, stream);
          break;

        case "$GPGBS" :
          new GBS(parts, stream);
          break;

        default:
          System.err.println("Unrecognised sentence : " + input);
      }
    }
  }


  /**
   * Synchronises the input from 2 streams depending on their timestamps
   * calls parseSentence() on earliest stream to advance it
   * or both if synced.
   * @param stream1 first Stream
   * @param stream2 second Stream
   */
  private void syncStream(Stream stream1, Stream stream2){
    boolean synced = false;

    while(!synced){
      checkForNewFix();
      if(stream1.getStreamTime().compareTo(stream2.getStreamTime()) < 0){
        parseSentence(stream1);
        synced = false;
      }else if (stream1.getStreamTime().compareTo(stream2.getStreamTime()) > 0){
        parseSentence(stream2);
        synced = false;
      }else{
        synced = true;
        calculateOffset(stream1, stream2);
        parseSentence(stream1);
        parseSentence(stream2);

      }
    }
  }

  /**
   * Checks streams for fix availability based on current
   * stream quality. Makes a new corrected GPS fix if
   * quality is sufficient favouring stream 1
   */
  private void checkForNewFix(){
    if(s1.getFixtype()== FixType.GOOD_FIX){
      makeGPSfix(s1);

    }else if(s2.getFixtype()== FixType.GOOD_FIX){
      makeGPSfix(s2);

    }else{
      if(s1.getFixtype()== FixType.MIN_FIX){
        makeGPSfix(s1);

      }else if(s2.getFixtype()== FixType.MIN_FIX){
        makeGPSfix(s2);

      }
    }
  }

  /**
   * Creates a corrected gps fix if 1 second has passed since the last corrected fix from
   * current data in the parameter Stream.
   *
   * @param stream the stream
   */
  private void makeGPSfix(Stream stream){
    if((lastFixTime == null) || (stream.getStreamTime().compareTo(lastFixTime) > 0)){
      gps.add(stream.makeGPS());
      updateFixTime(stream);

    }
  }

  /**
   * Updates the time of last corrected fix with time from the
   * stream which generated the fix.
   * @param stream the stream
   */
  private void updateFixTime(Stream stream){
    lastFixTime = (Calendar)stream.getStreamTime().clone();
  }

  /**
   * Calculates and applies an offset for for values in stream 2
   * @param stream1 the first stream
   * @param stream2 the second stream
   */
  private void calculateOffset(Stream stream1, Stream stream2){
    float latOffset = stream1.getCurrentLat() - stream2.getCurrentLat();
    float lngOffset = stream1.getCurrentLng() - stream2.getCurrentLng();
    float elevOffset = stream1.getCurrentElev() - stream2.getCurrentElev();

    stream2.updateLatLongOffest(latOffset, lngOffset);
    stream2.updateElevOffset(elevOffset);
  }


  /**
   * Creates a GPX file for corrected gps stream output
   * @param filename the filename
   */
  private void makeGPX(String filename){
     new GPXmaker(gps, filename);
  }



}

File: cpp/Controller.cpp
/**
 * @file   Controller.cpp
 * @Author Sion Griffiths (sig2@aber.ac.uk)
 * @date   March 2014
 * @brief  Controller sets out the central sequence of operations for the system
 * 
 *  
 */

#include "Controller.h"
#include "FileParser.h"
#include "GGA.h"
#include "RMC.h"
#include "GSV.h"
#include "GPXmaker.h"
#include <vector>
#include <string>
#include <string.h>
#include <sstream>
using namespace std;

/**
 * Constructs a controller, instantiates Stream objects 
 * using stream data files
 */
Controller::Controller() {
    lastFixTime = 0;
    s1 = new Stream("gps_1.dat");
    s2 = new Stream("gps_2.dat");
}

Controller::Controller(const Controller& orig) {
}

/**
 * Destructor, deletes Stream pointers
 */
Controller::~Controller() {
    delete s1;
    delete s2;
}

/**
 * Run method is the point of entry to the system.
 */
void Controller::run() {

    while (!s1->isEndOfStream()) {
        syncStreams(s1, s2);
    }
    cout << gpsfixes.size();
    new GPXmaker(gpsfixes);
}

/**
 * Tokenises the next sentence passed through Stream
 * from file. Creates a specific sentence type object
 * and passes through the tokenised sentence as vector
 * @param stream pointer to Stream
 */
void Controller::parseSentence(Stream *stream) {

    string input;
    if ((input = stream->getNext()).empty()) {
        stream->setEndOfStream(true);
        return;
    } else {

        vector<string> sentence = split(input);

        string type = sentence[0];
        if ((type.compare("$GPGGA")) == 0) {
            GGA(stream, sentence);
        } else if ((type.compare("$GPRMC")) == 0) {
            RMC(stream, sentence);
        } else if ((type.compare("$GPGSV")) == 0) {
            GSV(stream, sentence);
        }
    }
}

/**
 * Splits a comma separated sentence, returns sentence as 
 * a vector of tokenised string
 * @param s input string 
 * @return vector of tokens
 */
vector<string> Controller::split(const string &s) {

    vector <string> sentenceTokens;
    stringstream ss(s);

    while (ss.good()) {
        string substr;
        getline(ss, substr, ',');

        sentenceTokens.push_back(substr);
    }
    return sentenceTokens;
}

/**
 * Calculates the offset between 2 Streams and applies offset to values
 * in the second Stream
 * @param s1 pointer to first Stream
 * @param s2 pointer to second Stream
 */
void Controller::calculateOffsets(Stream* s1, Stream* s2) {
    float latOffset = s1->getCurrentLat() - s2->getCurrentLat();
    float lngOffset = s1->getCurrentLng() - s2->getCurrentLng();
    float elevOffset = s1->getCurrentElev() - s2->getCurrentElev();

    s2->updateLatLngOffsets(latOffset, lngOffset);
    s2->updateElevOffset(elevOffset);
}

/**
 * Synchronises the parsing of sentences from 2 Streams
 * using the current Stream times.
 * @param s1 pointer to first Stream
 * @param s2 pointer to second Stream
 */
void Controller::syncStreams(Stream *s1, Stream *s2) {

    bool synced = false;


    while (synced == false) {
        checkForNewFix();

        if (s1->getStreamTime() < (s2->getStreamTime())) {
            parseSentence(s1);
            synced = false;

        } else if (s1->getStreamTime() > (s2->getStreamTime())) {
            parseSentence(s2);
            synced = false;

        } else {

            calculateOffsets(s1, s2);
            parseSentence(s1);
            parseSentence(s2);
            synced = true;

        }
    }
}

/**
 * Checks if current state of streams is good enough
 * to make a new fix. Biased to Stream 1
 */
void Controller::checkForNewFix() {

    if (s1->currentQuality == s1->GOOD_FIX) {
        makeGPSfix(s1);

    } else if (s2->currentQuality == s2->GOOD_FIX) {
        makeGPSfix(s2);

    } else {
        if (s1->currentQuality == s1->MIN_FIX) {
            makeGPSfix(s1);

        } else if (s2->currentQuality == s2->MIN_FIX) {
            makeGPSfix(s2);

        }
    }
}

/**
 * Makes a new GPSfix from current stream state
 * @param stream pointer to Stream
 */
void Controller::makeGPSfix(Stream *stream) {
    if ((lastFixTime == 0) || (stream->getStreamTime() > lastFixTime)) {
        if(streamHasData(stream)){
            gpsfixes.push_back(stream->makeGPSfix());
            lastFixTime = stream->getStreamTime();
        }
    }
}

/**
 * Checks that a stream has sufficient data to make a fix
 * @param stream pointer to Stream
 * @return true if enough data, false otherwise.
 */
bool Controller::streamHasData(Stream *stream) {
    float elev = stream->getCurrentElev();
    float lat = stream->getCurrentLat();
    float lng = stream->getCurrentLng();

    return (lng != 0 && lat != 0 && elev != 0);
}